web/ui/react-app/src/pages/alerts/AlertContents.tsx

```diff
@@ -71,17 +72,22 @@ const AlertsContent: FC<AlertsProps> = ({ groups = [], statsCount }) => {
     inactive: true,
   });
   const [showAnnotations, setShowAnnotations] = useLocalStorage('alerts-annotations-status', { checked: false });
-
   const toggleFilter = (ruleState: RuleState) => () => {
     setFilter({
       ...filter,
       [ruleState]: !filter[ruleState],
     });
   };
 
-  const handleSearchChange = (e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
-    if (e.target.value !== '') {
-      const pattern = e.target.value.trim();
+  const updateURL = (expr: string): void => {
+    const query = encodeToQueryString(expr);
+    window.history.pushState({}, '', query);
+  };
```

Comment:
Given that the `updateURL` function is now repeated three times, how about merging the functionality with your current `encodeToQueryString()` utility function? And `encodeToQueryString()` also sounds a bit generic for a function that always modifies a *specific* query parameter, so maybe rename the resulting merged function to something like `setQuerySearchFilter()`, which does both the encoding and the URL modification?

And then thinking a bit further, it could also make sense to remove the duplication of the `useMemo` call where the query string is parsed. So in addition to a `setQuerySearchFilter()` utility function there could also be a `getQuerySearchFilter()` utility function that reads the parameter from the URL. And then the `useMemo` simply becomes `const defaultValue = useMemo(getQuerySearchFilter, [])` (after also removing the search change handler call, see other comment).

Reply:
ready! :)